src/bosh-director/lib/bosh/director/jobs/db_job.rb

```diff
@@ -51,26 +53,34 @@ def queue_name
       private
 
       def update_task_state
-        Config.db.transaction(retry_on: [Sequel::DatabaseConnectionError]) do
+        Config.db.transaction(wait: 1.seconds, attempts: 5,
+                              retry_on: [Sequel::DatabaseDisconnectError, Sequel::DatabaseConnectionError]) do
           task = Models::Task.where(id: @task_id).first
           raise DirectorError, "Task #{@task_id} not found in queue" unless task
 
           task.checkpoint_time = Time.now
-          if task.state == 'cancelling'
+          case task.state
+          when 'cancelling'
             task.state = 'cancelled'
             Config.logger.debug("Task #{@task_id} cancelled")
-          elsif task.state == 'queued'
+          when 'queued'
             task.state = 'processing'
           else
-            task.save
+            raise DirectorError, "Cannot update task state in DB for #{@task_id}" if task.save.nil?
+
             raise DirectorError, "Cannot perform job for task #{@task_id} (not in 'queued' state)"
+
           end
-          task.save
```

Comment:
Looking at the Sequel code, the database error can be thrown after the transaction `commit` has been issued, so the database successfully updates the record, but you still end up in the `retry` state here.

In that situation the `state` for the task after you find it from the database again will be `cancelling` or `processing` which will lead to the `else` case and a very strange error in the logs.

Reply:
I think I addressed that concern.